rust/src/lib/ovsdb/json_rpc.rs

```diff
@@ -37,44 +39,125 @@ pub(crate) struct OvsDbRpcReply {
         self.socket
             .write_all(buffer.as_bytes())
-            .map_err(parse_socket_io_error)?;
+            .await
+            .map_err(|e| {
+                NmstateError::new(
+                    ErrorKind::PluginFailure,
+                    format!("Failed to send message to OVSDB: {e}"),
+                )
+            })?;
+        self.socket.flush().await.map_err(|e| {
+            NmstateError::new(
+                ErrorKind::PluginFailure,
+                format!(
+                    "Failed to flush buffer when sending message to OVSDB: {e}"
+                ),
+            )
+        })?;
         Ok(())
     }
 
-    pub(crate) fn recv(
+    // * JSON-RPC has no indicator for `end-of-message`.
+    // * UnixStream has no indicator for `end-of-message`.
+    // * The OpenvSwitch code `lib/jsonrpc.c` function `jsonrpc_recv` is
+    //   depending on JSON parser to determine whether message ended, and keep
+    //   retry for `MAX_RECV_RETRY_COUNT` count.
+    pub(crate) async fn recv(
         &mut self,
         transaction_id: u64,
     ) -> Result<Value, NmstateError> {
-        let mut response: Vec<u8> = Vec::new();
-        loop {
-            let mut buffer = [0u8; BUFFER_SIZE];
-            let read = self
-                .socket
-                .read(&mut buffer)
-                .map_err(parse_socket_io_error)?;
-            log::debug!("OVSDB: recv data {:?}", &buffer[..read]);
-            response.extend_from_slice(&buffer[..read]);
-            if read < BUFFER_SIZE {
-                break;
+        let mut response: Vec<u8> = Vec::with_capacity(BUFFER_SIZE);
+
+        let mut reply: Result<OvsDbRpcReply, NmstateError> =
+            Err(NmstateError::new(
+                ErrorKind::PluginFailure,
+                "Empty reply from OVSDB".to_string(),
+            ));
+
+        for _ in 0..MAX_RECV_RETRY_COUNT {
+            self.socket.readable().await.map_err(|e| {
+                NmstateError::new(
+                    ErrorKind::PluginFailure,
+                    format!("OVSDB connection is not readable: {e}"),
+                )
+            })?;
+            let mut buffer = [0; BUFFER_SIZE];
+            let read_size = match self.socket.try_read(&mut buffer) {
+                Ok(s) => s,
+                Err(e) if e.kind() == std::io::ErrorKind::WouldBlock => {
+                    continue;
+                }
+                Err(e) => {
+                    return Err(NmstateError::new(
+                        ErrorKind::PluginFailure,
+                        format!(
+                            "Failed to read data from OVSDB connection: {e}"
+                        ),
+                    ))
+                }
+            };
+            log::debug!(
+                "OVSDB: recv size {read_size}, data {:?}",
+                &buffer[..read_size]
+            );
+            if read_size > 0 {
+                response.extend_from_slice(&buffer[..read_size]);
+            }
```

Comment:
This is one of the things that I dislike most of the current Rust: the inability to read to a buffer without initializing and copy lot of data. It would be much better to store it to the `Vec` directly.

Is it possible for us to use [try_read_buf](https://docs.rs/tokio/latest/tokio/net/struct.TcpStream.html#method.try_read_buf)? It ways that it needs the crate feature io-util. I guess we can't.

LGTM.

Reply:
Yes, but `Vec` also reallocs, and then you copy from `buffer` to `response`. Anyway, if there is no standard good way of doing it yet, let's leave as is.